### PR TITLE
Split different tests into separate Github Action jobs

### LINKED ISSUE #178: Run different tests as separate GitHub Action jobs
Currently the [test](https://github.com/localgovdrupal/drupal-container/blob/master/run-tests.sh) script runs coding standards and deprecated code checks and the PHP Unit tests, returning a failure if any one of these fails. This doesn't give developers much feedback on what failed. If each test was run as a separate job then it would be clear which of these failed.

### PR DESCRIPTION
**Pull request by stephen-cox**
See #178 for a description of the issue.

### CODE REVIEW

**Comment by paulpopus** [comment]
From a brief look at this test the issue might be that the subsequent job doesn't continue in the same directory as the previous one. Assuming that the environment doesn't change you can try to change dir to the one logged in the previous job **or** we can try to have a consistent working directory for every job.

**Comment by stephen-cox** [comment]
I finally have split out the tests so the coding standards, code deprecation and phpunit tests report separately, making it easier to see where the failure is.

If we're happy with this then I'll create a merge request for the template (https://github.com/localgovdrupal/.github/blob/master/workflow-templates/localgov-drupal-ci.yml) to use this approach and then create merge requests for all repos that run tests.

**Comment by finnlewis** [comment]
Looks good @stephen-cox  - it will be much easier to see when coding standards tests fail rather than PHPUnit tests. 

I'll let @Adnan-cds and @andybroomfield look, but seems logical and sensible to me.

**Comment by Adnan-cds** [comment]
> Do you have any thoughts on this approach to running tests?

Sorry, keep getting pushed back by this and that :(  I will definitely look into this tomorrow God willing :)

> it would be good to update the GitHub CI config for all modules either before or as part of the Drupal 9 update work.

Let's do it separately.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -1,54 +1,133 @@
...
-        run: docker-compose up -d
+        run: docker-compose -f docker-compose.yml up -d
 
-      - name: Run tests
-        run: ./run-tests.sh
-        shell: bash
+      - name: Run PHPUnit tests
+        run: |
+          mkdir -p ./html/web/sites/simpletest && chmod 777 ./html/web/sites/simpletest
+          sed -i "s#http://localgov.lndo.site#http://drupal#; s#mysql://database:database@database/database#sqlite://localhost/dev/shm/test.sqlite#" ./html/phpunit.xml.dist
```
Comment (Adnan-cds) [inline .github/workflows/test.yml:130]: A crucial slash may have disappeared from ```localhost//dev/shm/test.sqlite```.  Please compare this bash snippet with [run-test.sh](https://github.com/localgovdrupal/drupal-container/blob/master/run-tests.sh)

**Review by Adnan-cds — COMMENTED** [review]
I have added a few comments :)

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -1,54 +1,133 @@
...
+          composer global config github-oauth.github.com ${{ github.token }}
+
+      - name: Extract Git branch name outside of a pull request
+        if: github.event_name != 'pull_request'
+        run: echo "GIT_BRANCH=${GITHUB_REF#refs/heads/}" >> $GITHUB_ENV
 
-      - name: Save git branch and git repo names to env if is a pull request
+      - name: Extract Git branch name from a pull request
         if: github.event_name == 'pull_request'
+        run: echo "GIT_BRANCH=${GITHUB_HEAD_REF}" >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (stephen-cox) [reply .github/workflows/test.yml]: I have updated the build steps to include the recent changes to the template.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -1,54 +1,133 @@
...
+
+      - name: Grab test target
+        run: composer --working-dir=html require ${{ github.repository }}:"dev-${GIT_BRANCH} as 1.0.x-dev"
```
[the remark this answers is not in the document]
Comment (stephen-cox) [reply .github/workflows/test.yml]: I have added the --with-all-dependencies flag.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -1,54 +1,133 @@
...
+      
+    steps:
+
+      - name: Cached workspace
+        uses: actions/cache@v2
+        with:
+          path: ./html
+          key: ${{ runner.os }}-localgov-build-${{ github.run_id }}
+          restore-keys: |
+            ${{ runner.os }}-localgov-build-
```
[the remark this answers is not in the document]
Comment (stephen-cox) [reply .github/workflows/test.yml:70]: No typo here; `restore-keys` matches caches beginning with the string, so `${{ runner.os }}-localgov-build-` matches any cache beginning with that string.

This is just a fallback if the main cache indexed with the `run_id` is not found.

File: .github/workflows/test.yml (hunk 1 above)
Comment (stephen-cox) [reply .github/workflows/test.yml:134]: The extra slash is not actually needed. I discovered this while reading the Drupal docs on this: https://www.drupal.org/docs/automated-testing/phpunit-in-drupal/phpunit-browser-test-tutorial#s-sqlite

**Comment by stephen-cox** [comment]
@Adnan-cds Thanks for the feedback. I have made some changes and replied to your other comments. Tests are running

File: .github/workflows/test.yml (hunk 1 above)
Comment (Adnan-cds) [reply .github/workflows/test.yml:134]: Hi Steve, you correct that you can ignore the extra slash and the test will run just fine.  What happens with "sqlite://localhost/dev/shm/test.sqlite" is, a file called "dev/shm/test.sqlite" is created **inside** Drupal's docroot.  But if you use "sqlite://localhost//dev/shm/test.sqlite", you genuinely get /dev/shm/test.sqlite which speeds up the test run slightly.

I can see that even with "sqlite://localhost/dev/shm/test.sqlite", the PHPUnit tests have taken 16m47s which is not bad. This used to take about 14m in [previous test runs](https://github.com/localgovdrupal/localgov_alert_banner/runs/2468599165).  Not a lot of difference.  So I will leave this to you.

File: .github/workflows/test.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Adnan-cds) [reply .github/workflows/test.yml:70]: Thanks.

File: .github/workflows/test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Adnan-cds) [reply .github/workflows/test.yml]: Thanks.

File: .github/workflows/test.yml (hunk 1 above)
Comment (stephen-cox) [reply .github/workflows/test.yml]: @Adnan-cds  I've added the extra slash and the run took 14m19s. Nice catch.

Are you happy to approve this now?

**Review by Adnan-cds — APPROVED** [review]
Thanks a lot :)

**Pull request merged**